src/assign.c

```diff
@@ -200,6 +205,16 @@ static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
   return(newdt);
 }
 
+SEXP shallow(SEXP dt, SEXP cols) {
+  return _shallow(dt, cols, -1);
+}
+
+SEXP copy(SEXP x)
+{
+  return INHERITS(x, char_datatable) ? _shallow(x, R_NilValue, -1) : duplicate(x);
```

Comment:
This change is responsible for 10 failed tests. Previously, `copy()` only returned `duplicate(x)`.  An additional flag to guarantee copying or maybe even a different function may be helpful, e.g. `SEXP copy(SEXP x, bool doCopy) { return doCopy || !INHERITS(x, char_datatable) ? duplicate(x) : _shallow(x, R_NilValue, -1);}`

One way the tests fail is that the shallow copy returned does not always copy when assigned correctly when `i` is provided. E.g., `copy(dt)[, x := new_val]` would not affect the original dt whereas `copy(dt)[1L, x:= new_val]` would affect the original dt and it would be modified. These examples would be helpful to have a flag to guarantee a copy - here are the 8 tests that fail (980.20,  982.20, 1419.03, 1419.05, 1419.07, 1419.10, 2068.20, 2233.36).

A second way the tests failed is that now `[<-` behaves more similarly to `dt[, x:=newval]` in that other objects pointing to the same address gets similar change when the replacement vector is a different length (i.e., assigning a scalar to a column that is more than 1 value). There are 2 tests like this (104.2 and 113 )

Just as interesting to note is that when reverting `copy` back to simply `return duplicate(x);` there are 3 new errors produced that aren't produced in the original PR (510, 2016.6, 2049.4).  The first and last test are due to #5084 detecting invalid self ref. Specifically, making `<-` assignments and then subsequent assignment calls don't build on one another. e.g., `dt = data.table(x = 1L, y = 1L); key(dt) <- "x"; dt[, y:=2L];` results in `data.table(x = 1L, y = 1L)`. In that example, the first test would typically be correct but then the next text would expect the data.table to be modified but it is still the original.

2016.6, is also related to the invalid self ref, but the cause is because the data.table is loaded instead of modifying data.tables in a non-canonical way. I'll look at this more next weekend. 

```r
## Tests 980.20,  982.20, 1419.03, 1419.05, 1419.07, 1419.10, 2068.20, 2233.36
dt = data.table(x = 1L)
copy(dt)[, x:= x+1L]
dt
## x
## <int>
##   1:     1

copy(dt)[1L, x:=x+1L] ## affects original dt. Any subsequent calls that assume unmodified dt cause errors
dt
## x
## <int>
##   1:     2
```
``` r
## Tests 104.20  113.00
dt = data.table(a = 1:2)
dt1 <- dt
dt1[, "A"] <- 3L ## if this were the same length as initial vector, a copy would be made and match original tests
dt
##        a
##    <int>
## 1:     3
## 2:     3
```

```r
## Reverting copy behavior to always duplicate **change to PR code on local**
## Tests 510, 2049.4
dt = data.table(x = 1L, y = 1L)
key(dt) <- "x"
dt[, y:= 2L] ## Warning about invalid ref
dt           ## No change
## Key: <x>
##        x     y
##    <int> <int>
## 1:     1     1

## Similar issue when loading data from storage
## This still modified to return duplicate copy
## 2016.6
DT = data.table(a=1:3)
save(list="DT", file=tt<-tempfile())
rm(DT)
name = load(tt)
DT[, a:=4L] ## warning due to invalid self ref.
DT
##        a
##    <int>
## 1:     1
## 2:     2
## 3:     3

Reply:
Thanks for digging into this
> whereas `copy(dt)[1L, x:= new_val]` would affect the original dt and it would be modified.

This indeed seems to be unexpected behavior.

Reply:
added tests 492.* to ensure about that